pyclibrary/c_model.py

```diff
@@ -0,0 +1,779 @@
+
+The class hierarchy is:
+
+* CLibInterface (a catalogue of all CLibBase objects in a library)
+* CLibBase (base class of all objects in header file)
+  * Macro
+    * ValMacro (#define x ...)
+    * FnMacro (#define x(...) ...)
+  * CLibType
+    * SimpleType
+      * BuiltinType (int, char, double, ...)
+      * CustomType (all references to custom types = typedefs or named
+                    AlgebraicDataTypes)
+    * AlgebraicDataType
+      * EnumType
+      * CompoundType
+        * StructType
+        * UnionType
+        * BitFieldType
+    * ComposedType
+      * PointerType
+      * ArrayType
+      * FunctionType
+"""
+from __future__ import (division, unicode_literals, print_function,
+                        absolute_import)
+import collections
+import itertools
+import re
+from future.moves.itertools import zip_longest
+from .errors import UnknownCustomTypeError
+
+
+def _lpadded_str(text):
+    """An internal helper, returns '' if text is None otherwise ' '+text"""
+    if text is None:
+        return ''
+    else:
+        return ' ' + text
+
+
+class CLibBase(object):
+
+    __slots__ = ()
+
+    def _getattrnames(self):
+        """Internal method that lists all fields
+        :rtype: list[str]
+        """
+        for cls in type(self).__mro__:
+            if hasattr(cls, '__slots__'):
+                for name in cls.__slots__:
+                    yield name
+
+    def copy(self):
+        """Creates a shallow copy o this type.
+        :rtype: CLibBase
+        """
+        attrs = {anm: getattr(self, anm) for anm in self._getattrnames()}
+        return type(self)(**attrs)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        return all(getattr(self, name) == getattr(other, name)
+                   for cls in type(self).__mro__
+                   for name in getattr(cls, '__slots__', ()))
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __repr__(self):
+        code_obj = self.__init__.__code__
+        init_params = code_obj.co_varnames[1:code_obj.co_argcount]
+        default_vals = dict(zip(reversed(init_params),
+                                reversed(self.__init__.__defaults__ or [])))
+
+        def par_repr(param_name):
+            val = getattr(self, param_name)
+            if param_name not in default_vals:
+                yield repr(val)
+            else:
+                default_val = ([] if param_name == "quals"
+                               else default_vals[param_name])
+                if val != default_val:
+                    yield param_name + '=' + repr(val)
+
+        return (type(self).__name__ + '(' +
+                ', '.join(itertools.chain(*map(par_repr, init_params))) + ')')
+
+
+class CLibType(CLibBase):
+    """This is the abstract base class for all objects that are modeling a
+    "C" type.
+    """
+
+    __slots__ = ('quals',)
+
+    def __init__(self, quals=None):
```

Comment:
Nice. I read the some subclasses perhaps a bit to fast and missed the correspondance.